### PR TITLE
Implement forecast into the app.

### PR DESCRIPTION
**Pull request by prestosole**

### CODE REVIEW

File: packages/clima_data/lib/data_sources/forecast_remote_data_source.dart [hunk 1]
```diff
@@ -0,0 +1,51 @@
+import 'dart:convert';
+
+import 'package:clima_core/failure.dart';
+import 'package:clima_data/models/forcasts_model.dart';
+import 'package:clima_domain/entities/city.dart';
+import 'package:dartz/dartz.dart';
+import 'package:http/http.dart' as http;
+import 'package:riverpod/riverpod.dart';
+
+const String apiKey = '4bef3adf2fcb90307c2bf5feac75a2ba';
```
Comment (triallax) [inline packages/clima_data/lib/data_sources/forecast_remote_data_source.dart]: This should probably be put in a separate `constants.dart` file in `clima_data` or something.

File: packages/clima_data/lib/data_sources/forecast_remote_data_source.dart [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+abstract class ForecastsRemoteDataSource {
+  Future<Either<Failure, ForecastsModel>> getWeather(City city);
```
Comment (triallax) [inline packages/clima_data/lib/data_sources/forecast_remote_data_source.dart]: This should be renamed to `getForecasts`.

File: packages/clima_domain/lib/entities/forecasts.dart [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import 'package:clima_domain/entities/weather.dart';
+import 'package:meta/meta.dart';
+
+class Forecasts with Equatable<Forecasts> {
+  Forecasts({
+    @required this.forecasts,
+  });
+  final List<Weather> forecasts;
+  @override
+  // TODO: implement checkRuntimeType
```
Comment (triallax) [inline packages/clima_domain/lib/entities/forecasts.dart]: You probably forgot to remove this comment and the one below.

File: packages/clima_domain/lib/repos/forecasts_repo.dart [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+import 'package:clima_domain/entities/city.dart';
+import 'package:clima_domain/entities/forecasts.dart';
+import 'package:dartz/dartz.dart';
+import 'package:riverpod/riverpod.dart';
+
+abstract class ForcastsRepo {
+  Future<Either<Failure, Forecasts>> getForcasts(City city);
+}
+
+final weatherRepoProvider =
```
Comment (triallax) [inline packages/clima_domain/lib/repos/forecasts_repo.dart]: This should be renamed to `forecastsRepoProvider`. It should also be overriden with `forecastsRepoImplProvider` in `clima_ui/lib/main.dart` (see https://github.com/CentaurusApps/clima/blob/master/packages/clima_ui/lib/main.dart#L20 for an example).

**Review by triallax — CHANGES_REQUESTED** [review]
Mostly fine; you should add tests too.

**Comment by triallax** [comment]
There seems to be some extraneous commits. Can you try to rebase?

File: packages/clima_data/lib/models/forcasts_model.dart [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  ForecastsModel({
+    @required List<Weather> forecasts,
+    @required String cityName,
+  }) : super(forecasts: forecasts, cityName: cityName);
+
+  factory ForecastsModel.fromJson(Map<String, dynamic> json) {
+    final list = json['list'] as List<dynamic>;
+    return ForecastsModel(
+      forecasts: list.map((e) => ForecastModel.fromJson(e)).toList(),
+    );
```
Comment (triallax) [inline packages/clima_data/lib/models/forcasts_model.dart]:
```suggestion
    final cityName = json['city']['name'] as String;

    return ForecastsModel(
      cityName: cityName,
      forecasts: list.map((e) => ForecastModel.fromJson(e, cityName: cityName)).toList(),
    );
```

File: packages/clima_data/lib/models/forecast_model.dart [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+      windSpeed: (json['wind']['speed'] as num).toDouble() * 3.6,
+      condition: json['weather'][0]['id'] as int,
+      cityName: json['city']['name'] as String,
+      description: json['weather'][0]['description'] as String,
+      time: (json['timezone'] as num).toInt(),
+      iconCode: json['weather'][0]['icon'] as String,
+      sunrise: json['sys']['sunrise'] as int,
+      sunset: json['sys']['sunset'] as int,
+      humidity: json['main']['humidity'] as int);
+}
```
Comment (triallax) [inline packages/clima_data/lib/models/forecast_model.dart]:
```suggestion
  factory ForecastModel.fromJson(Map<String, dynamic> json, {@required String cityName}) => ForecastModel(
      temperature: (json['main']['temp'] as num).toDouble(),
      tempMax: (json['main']['temp_max'] as num).toDouble(),
      tempMin: (json['main']['temp_min'] as num).toDouble(),
      tempFeel: (json['main']['feels_like'] as num).toDouble(),
      // We multiply by 3.6 to convert from m/s to km/h.
      windSpeed: (json['wind']['speed'] as num).toDouble() * 3.6,
      condition: json['weather'][0]['id'] as int,
      cityName: cityName,
      description: json['weather'][0]['description'] as String,
      time: (json['timezone'] as num).toInt(),
      iconCode: json['weather'][0]['icon'] as String,
      sunrise: json['sys']['sunrise'] as int,
      sunset: json['sys']['sunset'] as int,
      humidity: json['main']['humidity'] as int);
}
```

File: packages/clima_domain/lib/entities/forecasts.dart [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    @required this.cityName,
+  });
+  final List<Weather> forecasts;
+  final String cityName;
+  @override
+  bool get checkRuntimeType => false;
+
+  @override
+  List<Object> get props => [
+        forecasts,
```
Comment (triallax) [inline packages/clima_domain/lib/entities/forecasts.dart:17]: You forgot to add `cityName` here.

File: packages/clima_data/lib/repos/forecasts_repo_impl.dart [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  @override
+  Future<Either<Failure, Forecasts>> getForecasts(City city) async {
+    if (await connectivity.checkConnectivity() == ConnectivityResult.none) {
+      return Left(const NoConnection());
+    } else {
+      final memoizedForecasts = await memoizedDataSource.getMemoizedForecasts();
+
+      if (memoizedForecasts.isLeft() ||
+          memoizedForecasts.all(
+              (weather) => weather != null && weather.cityName == city.name)) {
```
Comment (triallax) [inline packages/clima_data/lib/repos/forecasts_repo_impl.dart]:
```suggestion
              (forecasts) => forecasts != null && forecasts.cityName == city.name)) {
```

File: packages/clima_domain/lib/entities/forecasts.dart (hunk 7 above)
Comment (prestosole) [reply packages/clima_domain/lib/entities/forecasts.dart:17]: Oopsie!

File: packages/clima_domain/lib/use_cases/get_forecasts.dart [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+import 'package:clima_core/use_case.dart';
+import 'package:clima_domain/entities/city.dart';
+import 'package:clima_domain/entities/forecasts.dart';
+import 'package:clima_domain/repos/forecasts_repo.dart';
+import 'package:dartz/dartz.dart';
+import 'package:equatable/equatable.dart';
+import 'package:meta/meta.dart';
+import 'package:riverpod/riverpod.dart';
+
+class GetWeather implements UseCase<Forecasts, GetWeatherParams> {
```
Comment (triallax) [inline packages/clima_domain/lib/use_cases/get_forecasts.dart:11]: This should be `GetForecasts`.

**Review by triallax — CHANGES_REQUESTED** [review]

File: packages/clima_domain/lib/use_cases/get_forecasts.dart [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+  final ForecastsRepo repo;
+
+  const GetWeather(this.repo);
+
+  @override
+  Future<Either<Failure, Forecasts>> call(GetWeatherParams params) =>
+      repo.getForecasts(params.city);
+}
+
+class GetWeatherParams extends Equatable {
```
Comment (triallax) [inline packages/clima_domain/lib/use_cases/get_forecasts.dart:21]: And this should be `GetForecastsParams`.

File: packages/clima_domain/lib/use_cases/get_forecasts.dart [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+  const GetWeatherParams({@required this.city});
+
+  final City city;
+
+  @override
+  List<Object> get props => [city];
+}
+
+final getWeatherProvider =
```
Comment (triallax) [inline packages/clima_domain/lib/use_cases/get_forecasts.dart:30]: This should be `getForecastsProvider`.

**Review by triallax — APPROVED** [review]

**Pull request merged**